crates/matrix-sdk/src/room/futures.rs

```diff
@@ -319,3 +335,188 @@ impl<'a> IntoFuture for SendAttachment<'a> {
+    tracing_span: Span,
+    request_config: Option<RequestConfig>,
+}
+
+#[cfg(feature = "experimental-encrypted-state-events")]
+impl<'a> SendRawStateEvent<'a> {
+    pub(crate) fn new(
+        room: &'a Room,
+        event_type: &'a str,
+        state_key: &'a str,
+        content: impl IntoRawStateEventContent,
+    ) -> Self {
+        let content = content.into_raw_state_event_content();
+        Self {
+            room,
+            event_type,
+            state_key,
+            content,
+            tracing_span: Span::current(),
+            request_config: None,
+        }
+    }
+
+    /// Assign a given [`RequestConfig`] to configure how this request should
+    /// behave with respect to the network.
+    pub fn with_request_config(mut self, request_config: RequestConfig) -> Self {
+        self.request_config = Some(request_config);
+        self
+    }
+
+    /// Determines whether the inner state event should be encrypted before
+    /// sending.
+    ///
+    /// This method checks two conditions:
+    /// 1. Whether the room supports encrypted state events, by inspecting the
+    ///    room's encryption state.
+    /// 2. Whether the event type is considered "critical" or excluded from
+    ///    encryption under MSC3414.
+    ///
+    /// # Returns
+    ///
+    /// Returns `true` if the event should be encrypted, otherwise returns
+    /// `false`.
+    fn should_encrypt(room: &Room, event_type: &str) -> bool {
+        if !room.encryption_state().is_state_encrypted() {
+            trace!("Sending plaintext event as the room does NOT support encrypted state events.");
+            return false;
+        }
+
+        // Check the event is not critical.
+        if matches!(
+            event_type,
+            "m.room.create"
+                | "m.room.member"
+                | "m.room.join_rules"
+                | "m.room.power_levels"
+                | "m.room.third_party_invite"
+                | "m.room.history_visibility"
+                | "m.room.guest_access"
+                | "m.room.encryption"
+                | "m.space.child"
+                | "m.space.parent"
+        ) {
+            trace!("Sending plaintext event as its type is excluded from encryption.");
+            return false;
+        }
+
+        true
+    }
+}
+
+#[cfg(feature = "experimental-encrypted-state-events")]
+impl<'a> IntoFuture for SendRawStateEvent<'a> {
+    type Output = Result<send_state_event::v3::Response>;
+    boxed_into_future!(extra_bounds: 'a);
+
+    fn into_future(self) -> Self::IntoFuture {
+        let Self { room, mut event_type, state_key, mut content, tracing_span, request_config } =
+            self;
+
+        let fut = async move {
+            room.ensure_room_joined()?;
+
+            let mut state_key = state_key.to_owned();
+
+            if Self::should_encrypt(room, event_type) {
+                use tracing::debug;
+
+                Span::current().record("is_room_encrypted", true);
```

Comment:
the name of this field is slightly misleading... maybe:

```suggestion
                Span::current().record("should_encrypt", true);
```

... and likewise below for the `false` case`.

Reply:
I don't necessarily think `SendRawMessageLikeEvent` needs to match, tbh.